docs: all intention documentation updates (#8869)
hashicorp/consul: website/pages/docs/agent/config-entries/service-intentions.mdx

---
layout: docs
page_title: 'Configuration Entry Kind: Service Intentions'
sidebar_title: service-intentions <sup>Beta</sup>
description: >-
  The service-intentions config entry kind controls Connect traffic

>   authorization for both networking layer 4 (e.g. TCP) and networking layer 7

Reviewer: Should we name the layers? Our marketing material will use phrases like "application layer" vs "transport layer", so this could be another place to improve both comprehension and SEO.

Author: I was saving the marketing terminology for a later pass after I get all of the main parts squared away.

  (e.g. HTTP).
---

# Service Intentions <sup>Beta</sup>

-> **1.9.0+:** This config entry is available in Consul versions 1.9.0 and newer.

The `service-intentions` config entry kind controls Connect traffic

> authorization for both networking layer 4 (e.g. TCP) and networking layer 7

Reviewer: (Same here as the comment above)

```hcl
Kind = "service-router"
Name = "web"
Routes = [
  {
    Match {
      HTTP {
        Header = [
          {
            Name  = "x-debug"
            Exact = "1"
          },
        ]
      }
    }
    Destination {
      Service       = "web"
      ServiceSubset = "canary"
    }
  },
  {
    Match {
      HTTP {
        QueryParam = [
          {
            Name  = "x-debug"
            Exact = "1"
          },
        ]
      }
    }
    Destination {
      Service       = "web"
      ServiceSubset = "canary"
    }
  },
  # NOTE: a default catch-all will send unmatched traffic to "web"
]
```

Re-route a gRPC method to another service. Since gRPC method calls [are
HTTP2](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md), we can use an HTTP path match rule to re-route traffic:

```hcl
Kind = "service-router"
Name = "billing"
Routes = [
  {
    Match {
      HTTP {
        PathExact = "/mycompany.BillingService/GenerateInvoice"
      }
    }

    Destination {
      Service = "invoice-generator"
    }
  },
  # NOTE: a default catch-all will send unmatched traffic to "billing"
]
```

## Available Fields
TODO

TODO

> Permission precedence is applied top to bottom.

Reviewer: This sentence seems out of place.

  config entry migration and is primarily used to allow legacy intention
  [API](/api-docs/connect/intentions#update-intention-by-id)
  [endpoints](/api-docs/connect/intentions#read-specific-intention-by-id) to
  continue to function for a period of time after ugprading to 1.9.0.
  TODO(link-do-long-form-docs)

- `LegacyUpdateTime` `(time: optional)` - The timestamp that this intention was
  last updated. Cannot be set directly and is exposed here as an artifact of
  the config entry migration and is primarily used to allow legacy intention
  [API](/api-docs/connect/intentions#update-intention-by-id)
  [endpoints](/api-docs/connect/intentions#read-specific-intention-by-id) to
  continue to function for a period of time after ugprading to 1.9.0.
  TODO(link-do-long-form-docs)

### `IntentionPermission`

- `Action` `(string: <required>)` - This is one of "allow" or "deny" for
  the action that should be taken if this permission matches a request.

- `HTTP` `(IntentionHTTPPermission: <required>)` - A set of
  [http-specific authorization criteria](#intentionhttppermission).

### `IntentionHTTPPermission`

- `PathExact` `(string: "")` - Exact path to match on the HTTP request path.

>   At most only one of `PathExact`, `PathPrefix`, or `PathRegex` may be configured.

Reviewer: Do we usually include input field requirements/restrictions here? i.e. accepted characters?

Author: Yes, and these have no restrictions that consul enforces currently other than "PathExact/PathPrefix must start with a leading slash", just like service-router. Practically speaking envoy is going to complain if these make no sense at that layer, but they're largely just opaque string fields to Consul.

Reviewer: Oh that's an interesting requirement. We could add the slash for the user if they forget it (in a future iteration of this project) cc @jnwright

Reviewer 2: John has already added an error message if users forget to add the slash 🎉

- `PathPrefix` `(string: "")` - Path prefix to match on the HTTP request path.

  At most only one of `PathExact`, `PathPrefix`, or `PathRegex` may be configured.

- `PathRegex` `(string: "")` - Regular expression to match on the HTTP
  request path.

  The syntax is [described below](#regular-expression-syntax).

  At most only one of `PathExact`, `PathPrefix`, or `PathRegex` may be configured.

- `Methods` `(array<string>)` - A list of HTTP methods for which this match

>   applies. If unspecified all http methods are matched.

Reviewer: Rather than "If unspecified", would it be clearer to say "If empty"

Reviewer: Hm, I guess that's different than "unspecified". Maybe nevermind.

---
layout: docs
page_title: 'Configuration Entry Kind: Service Intentions (beta)'
sidebar_title: service-intentions <sup>Beta</sup>
description: >-
  The service-intentions config entry kind controls Connect traffic
  authorization for both networking layer 4 (e.g. TCP) and networking layer 7
  (e.g. HTTP).
---

> # Service Intentions <sup>Beta</sup>

Reviewer 3: ```suggestion
# Service Intentions ((#service-intentions)) <sup>Beta</sup>
```
I didn't find anything which currently links to this. Updating it for consistency with other beta heading names (per these suggestions).

  this reverse sorted by intention precedence, as that is the order that they
  will be evaluated at enforcement time.

### `SourceIntention`

- `Name` `(string: <required>)` - The source of the intention.
  For a `Type` of `consul` this is the name of a Consul service. The service
  doesn't need to be registered.

- `Namespace` `(string: "default")` <EnterpriseAlert inline /> - The namespace
  for the `Name` parameter.

- `Action` `(string: "")` - For an L4 intention this is required, and should be
  set to one of `"allow"` or `"deny"` for the action that should be taken if
  this intention matches a request.

  This should be omiitted for an L7 intention as it is mutually exclusive with
  the `Permissions` field.

- `Permissions` `(array<IntentionPermission>)` - The list of all [additional L7
  attributes](#intentionpermission) that extend the intention match criteria.

  Permission precedence is applied top to bottom. For any given request the
  first permission to match in the list is terminal and stops further
  evaluation. As with L4 intentions, traffic that fails to match any of the
  provided permissions in this intention will be subject to the default

>   intention/acl policy.

Reviewer 3: ```suggestion
  intention/ACL policy.
```

Author: I resolved this by actually cribbing a section from the older intentions docs that had a much better blurb.

>   policy](https://learn.hashicorp.com/tutorials/consul/access-control-setup-production).

Author: Fixed this and the place i copied this from over on the main intentions doc.

- `LegacyCreateTime` `(time: optional)` - The timestamp that this intention was
  created. Cannot be set directly and is exposed here as an artifact of the
  config entry migration and is primarily used to allow legacy intention
  [API](/api-docs/connect/intentions#update-intention-by-id)
  [endpoints](/api-docs/connect/intentions#read-specific-intention-by-id) to
  continue to function for a period of time after ugprading to 1.9.0.
  TODO(link-do-long-form-docs)

- `LegacyUpdateTime` `(time: optional)` - The timestamp that this intention was
  last updated. Cannot be set directly and is exposed here as an artifact of
  the config entry migration and is primarily used to allow legacy intention
  [API](/api-docs/connect/intentions#update-intention-by-id)
  [endpoints](/api-docs/connect/intentions#read-specific-intention-by-id) to
  continue to function for a period of time after ugprading to 1.9.0.
  TODO(link-do-long-form-docs)

### `IntentionPermission`

- `Action` `(string: <required>)` - This is one of "allow" or "deny" for
  the action that should be taken if this permission matches a request.

- `HTTP` `(IntentionHTTPPermission: <required>)` - A set of

>   [http-specific authorization criteria](#intentionhttppermission).

Reviewer 3: ```suggestion
  [HTTP-specific authorization criteria](#intentionhttppermission).
```

### `IntentionHTTPPermission`

- `PathExact` `(string: "")` - Exact path to match on the HTTP request path.

  At most only one of `PathExact`, `PathPrefix`, or `PathRegex` may be configured.

- `PathPrefix` `(string: "")` - Path prefix to match on the HTTP request path.

  At most only one of `PathExact`, `PathPrefix`, or `PathRegex` may be configured.

- `PathRegex` `(string: "")` - Regular expression to match on the HTTP
  request path.

  The syntax is [described below](#regular-expression-syntax).

  At most only one of `PathExact`, `PathPrefix`, or `PathRegex` may be configured.

- `Methods` `(array<string>)` - A list of HTTP methods for which this match

>   applies. If unspecified all http methods are matched.
>   If provided the names must be [valid](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods).

Reviewer 3: ```suggestion
  applies. If unspecified all HTTP methods are matched.
  If provided the names must be a valid [method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods).
```

    present with any value.

    At most only one of `Exact`, `Prefix`, `Suffix`, `Regex`, or `Present`
    may be configured.

  - `Exact` `(string: "")` - Match if the header with the given name is this
    value.

    At most only one of `Exact`, `Prefix`, `Suffix`, `Regex`, or `Present`
    may be configured.

  - `Prefix` `(string: "")` - Match if the header with the given name has
    this prefix.

    At most only one of `Exact`, `Prefix`, `Suffix`, `Regex`, or `Present`
    may be configured.

  - `Suffix` `(string: "")` - Match if the header with the given name has
    this suffix.

    At most only one of `Exact`, `Prefix`, `Suffix`, `Regex`, or `Present`
    may be configured.

  - `Regex` `(string: "")` - Match if the header with the given name matches
    this pattern.

    The syntax is [described below](#regular-expression-syntax).

    At most only one of `Exact`, `Prefix`, `Suffix`, `Regex`, or `Present`
    may be configured.

  - `Invert` `(bool: false)` - Inverts the logic of the match.

## ACLs

Configuration entries may be protected by
[ACLs](https://learn.hashicorp.com/tutorials/consul/access-control-setup-production).

> Reading a `service-intentions` config entry requires `intentions:read` on itself.

Reviewer 3: ```suggestion
Reading a `service-intentions` config entry requires `intentions:read` on the destination service.
```

Reviewer 3: If you agree with this suggestion, should we also consider changing this on the docs for the other L7 config entries?

Author: How about just the more generic "on its name", since that is unambiguous and the same across all of them.

Author: :+1:

> [ACLs](https://learn.hashicorp.com/tutorials/consul/access-control-setup-production).

Author: Updated on this page and all other kind pages.

Creating, updating, or deleting a `service-intentions` config entry requires

> `intentions:write` on itself.

Reviewer 3: ```suggestion
`intentions:write` on the destination service.
```
